### ISSUE TITLE
Makes `t()` available on context via a provider, instead of on props via modified connect

### ISSUE THREAD

**Issue by danjm** [MetaMask/metamask-extension#3787]
tldr: with addition of source maps, `metamask-connect.js` is broken in some cases (meaning it was a bad solution in the first place), need to place our i18n method `t` on context instead (which ends up being a better approach anyway, and will be strongly supported with react 16.3)

----

Currently on master, but not yet in prod, we make our i18n method - called `t` in our components - available on all components props via a wrapper around the `react-redux` `connect()` method. This is included in a filed called `metamask-connect.js`

The goals of this were to support the addition of adding localeMessages to redux (which in turn gives us improved performance, reliability, ease of updating/managing for users and developers) while also
- minimizing the additions a developer has to make to a component to get translations working
- minimizing the potential errors that could be caused by not passing the correct data to the i18n helper method

This seemed to be a sound approach given the react-redux api and the widespread use of higher order functions / components in react and redux development. The approach also worked when testing it in development.

However, with the addition of source maps, rebuild metamask reveals a bug in wrapping `connect()`. Specifically: connect() handles `mapStateToProps` different depending on how many arguments it has. It does seem safe to have the `metamask-connect` wrapper provide `connect()` a mapStateToProps with the same function signature every time (`function (state, ownProps = {})`). However, compilation to es5 results creates a function only having 1 parameter `state`, with `ownProps` being defined with a reference to `arguments`. In certain cases, due to the internals of `connect`, this causes `ownProps` to be empty when it should not be. Alternatives where our modified `mapStateToProps` has a function signature of 1 (`function (state)`) are not compatible with react-redux, which prevents modifications of `ownProps`.

A better solution will be to make the t method available via context by providing it through a wrapper component around the `SelectedApp` component in `root.js`

I had avoided this solution before because official react docs warn this the context api may stop be supported at any time. (Although, it is used by many widely used react libraries, e.g. `react-router` However, recent developments alleviate this concern, as react 16.3 will provide a new context API that is already being widely celebrated by the react community (see https://medium.com/dailyjs/reacts-%EF%B8%8F-new-context-api-70c9fe01596b, which incidentally shows how it can be used for a `LanguageProvider`).

I have tested the app with version 16.2 before, and it works without any apparent complications. We can use the current context api until react 16.3 comes out of beta, and then move over. We should be able to contain this migration to a single component.

Beyond fixing currently broken cases, using context further achieves the earlier stated goals by lowering the amount of changes/additions a component needs to support a redux connected i18n method.

To do this, we need to do the following:
- create an `I18nProvider` component that is connected to state and makes the `t` method bound to `localeMessages` available on the contexts of its children
- revert all uses of `metamask-connect` to `(`react-redux`).connect`
- add `t` to the context types of all components (This may seem like an increase in developer work per component, but eventually we will be adding props type validation to all components anyway, so whether the `t` method is validated in `propTypes` or `contextTypes` is irrelevant)
- replace all references to `this.props.t` with `this.context.t`

That should be sufficient to get everything working.

**Comment by kumavis** [comment]
Good research 👏👏
